Storage/tests/System/KmsTest.php

```diff
@@ -0,0 +1,279 @@
+        $keyNames[] = self::getCryptoKeyName($wrapper, $projectId, $keyRingId, $cryptoKeyId2);
+
+        return $keyNames;
+    }
+
+    /**
+     * @param RequestWrapper $wrapper
+     * @param string $projectId
+     * @param string $keyRingId
+     */
+    private static function buildKeyRing(
+        RequestWrapper $wrapper,
+        $projectId,
+        $keyRingId
+    ) {
+        try {
+            $wrapper->send(
+                new Request(
+                    'POST',
+                    sprintf(
+                        'https://cloudkms.googleapis.com/v1/projects/%s/locations/global/keyRings?keyRingId=%s',
+                        $projectId,
+                        $keyRingId
+                    )
+                )
+            );
+        } catch (ConflictException $ex) {
+            // If it already exists, great!
+        }
+    }
+
+    /**
+     * @param RequestWrapper $wrapper
+     * @param string $projectId
+     * @param string $keyRingId
+     * @param string $cryptoKeyId
+     * @return string
+     */
+    private static function getCryptoKeyName(
+        RequestWrapper $wrapper,
+        $projectId,
+        $keyRingId,
+        $cryptoKeyId
+    ) {
+        $name = null;
+
+        try {
+            $response = $wrapper->send(
+                new Request(
+                    'POST',
+                    sprintf(
+                        'https://cloudkms.googleapis.com/v1/projects/%s/locations/global/keyRings/%s/cryptoKeys?cryptoKeyId=%s',
+                        $projectId,
+                        $keyRingId,
+                        $cryptoKeyId
+                    ),
+                    [],
+                    json_encode(['purpose' => 'ENCRYPT_DECRYPT'])
+                )
+            );
+
+            $name = json_decode((string) $response->getBody(), true)['name'];
+        } catch (ConflictException $ex) {
+            $name = sprintf(
+                'projects/%s/locations/global/keyRings/%s/cryptoKeys/%s',
+                $projectId,
+                $keyRingId,
+                $cryptoKeyId
+            );
+        }
+
+        $policy = [
+            'policy' => [
+                'bindings' => [
+                    [
+                        'role' => 'roles/cloudkms.cryptoKeyEncrypterDecrypter',
+                        'members' => [
+                            "serviceAccount:$projectId@gs-project-accounts.iam.gserviceaccount.com"
+                        ]
+                    ]
+                ]
+            ]
+        ];
+        $wrapper->send(
+            new Request(
+                'POST',
+                sprintf(
+                    'https://cloudkms.googleapis.com/v1/projects/%s/locations/global/keyRings/%s/cryptoKeys/%s:setIamPolicy',
+                    $projectId,
+                    $keyRingId,
+                    $cryptoKeyId
+                ),
+                [],
+                json_encode($policy)
+            )
+        );
```

Comment:
This isn't a test, just a helper to set up the keys. If the policy isn't set as expected the rest of the expectations will fail, does that seem okay?

Reply:
That is correct.

Reply:
Potentially, you can still set the defaultKmsKeyName without the permission, but tests should fail on verifying an object insert with the bucket default kms key as it will not be able to encrypt the object.